File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-7.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charPos = {}
        res = 0
        start = 0
        for i in range(len(s)):
            print(s[i])
            if s[i] in charPos:
                res = max(res, i - charPos[s[i]])
                start = charPos[s[i]] + 1
                charPos[s[i]] = i
            else :
                res = max(res, i - start)
                charPos[s[i]] = i
        return res

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-7.cpp
class Solution {
private:
    std::unordered_map<std::pair<int, int>, int> cache;
public:
    int lengthOfLIS(vector<int>& nums) {
        return dp (nums, 0, 0);
    }
    int dp(vector<int>& nums, int i, int start) {
        if (i == nums.size()){
            return 0;
        }
        if (cache.find({i, start}) != cache.end()) {
            return cache[{i, start}];
        }
        int res = 0;
        for (int j = i + 1; j < nums.size(); ++j) {
            if (nums[i] < nums[j])
                res = std::max(res, 1 + dp(nums, j, start));
        } 
        res = std::max(res, 1 + dp(nums, nums.size(), start));
        res = std::max(res, dp(nums, i + 1, i + 1));
        cache[{i, start}] = res;
        return res;
    }
};
